lib/firmware/fsm_msg_nano.h

```diff
@@ -0,0 +1,309 @@
+
+    // Validate input data
+    bool invalid = false;
+    if (msg->has_parent_block) {
+        invalid |= msg->parent_block.has_parent_hash
+            && msg->parent_block.parent_hash.size != 32;
+        invalid |= msg->parent_block.has_link
+            && msg->parent_block.link.size != 32;
+        invalid |= msg->parent_block.balance.size != 16;
+    }
+    if (!msg->has_parent_block) {
+        invalid |= !msg->has_link_hash; // first block must be a receive block
+    }
+    invalid |= (int)msg->has_link_hash + (int)msg->has_link_recipient + (int)(msg->link_recipient_n_count != 0) > 1;
+    invalid |= msg->has_link_hash && msg->link_hash.size != 32;
+    invalid |= msg->balance.size != 16;
+    if (invalid) {
+        fsm_sendFailure(FailureType_Failure_Other, _("Block data invalid"));
+        layoutHome();
+        return;
+    }
+
+    bignum256 parent_balance;
+    bignum256 balance;
+    bignum256 balance_delta;
+    bn_zero(&parent_balance);
+    bn_zero(&balance);
+    bn_zero(&balance_delta);
+
+    uint8_t block_hash[32];
+    uint8_t parent_hash[32];
+    uint8_t link[32];
+    ed25519_public_key representative_pk;
+    uint8_t balance_be[16];
+
+    memset(block_hash, 0, sizeof(block_hash));
+    memset(parent_hash, 0, sizeof(parent_hash));
+    memset(link, 0, sizeof(link));
+    memset(representative_pk, 0, sizeof(representative_pk));
+    memset(balance_be, 0, sizeof(balance_be));
+
+    // Calculate parent hash
+    if (msg->has_parent_block) {
+        if (msg->parent_block.has_parent_hash) {
+            memcpy(parent_hash, msg->parent_block.parent_hash.bytes, sizeof(parent_hash));
+        }
+        memcpy(link, msg->parent_block.link.bytes, sizeof(link));
+        invalid |= !nano_validate_address(
+            coin->nanoaddr_prefix, strlen(coin->nanoaddr_prefix),
+            msg->parent_block.representative, strlen(msg->parent_block.representative),
+            representative_pk);
+        memcpy(balance_be, msg->parent_block.balance.bytes, sizeof(balance_be));
+        bn_from_bytes(balance_be, sizeof(balance_be), &parent_balance);
+
+        nano_hash_block_data(account_pk, parent_hash, link,
+                             representative_pk, balance_be,
+                             block_hash);
+
+        memcpy(parent_hash, block_hash, sizeof(parent_hash));
+        memset(block_hash, 0, sizeof(parent_hash));
+        memset(link, 0, sizeof(link));
+        memset(representative_pk, 0, sizeof(representative_pk));
+        memset(balance_be, 0, sizeof(balance_be));
+    }
+
+    // Calculate current hash
+    if (msg->has_link_hash) {
+        memcpy(link, msg->link_hash.bytes, sizeof(link));
+    } else if (msg->link_recipient_n_count > 0) {
+        node = fsm_getDerivedNode(coin->curve_name, msg->link_recipient_n, msg->link_recipient_n_count, NULL);
+        if (!node) return;
+        hdnode_fill_public_key(node);
+        memcpy(link, &node->public_key[1], sizeof(link));
```

Comment:
Should add a `memzero(node, sizeof(*node));` here to clear the secret from that static. If the block were invalid, we'd have left the secrets in it.

Reply:
Should the `node` content be zeroed at the end of the function as well then? or is that handled somewhere else?

Reply:
Yeah, it should be zeroed out when no longer needed.

Reply:
Addressed in fe8cfee